File: part2-object-oriented-prog/inheritance.js
"use strict"
class Animal {
    constructor(name) {
        this._name = name
    }

    speak() {
        console.log("Generic animals are Mute")
    }

    get name() {
        return this._name;
    }
}

class Dog extends Animal {

    constructor(name) {
        super(name);
        this.numLegs = 4;
    }

    speak() {
        console.log("woof woof");
    }
}

class CartoonDog extends Dog {

    speak() {
        console.log("I'll have a dry martini")
    }

}

var spot = new Dog('spot');
spot.speak();
console.log(spot.numLegs);
var brian = new CartoonDog('Brian')
brian.speak()
console.log(brian.name)

File: part2-object-oriented-prog/inheritance.py
class Animal:

    def __init__(self,name):
        self._name = name

    def speak(self):
        print("Generic Animals are mute")

    def get_name(self):
        return "My name is {}".format(self._name)

    name = property(get_name)


class Dog(Animal):
    def __init__(self,name):
        Animal.__init__(self,name) # super.__init__(name) in Python 3
        self.numLegs = 4


    def speak(self):
        print("woof woof")

class CartoonDog(Dog):

    def speak(self):
        print("I'll have a dry martini.")

spot = Dog("spot")
print(spot.name)
spot.speak()

brian = CartoonDog("Brian")
brian.speak()
print(brian.name)
